oeq-ts-rest-api/src/AxiosInstance.ts

```diff
@@ -17,17 +17,19 @@ const catchHandler = (error: AxiosError | Error): never => {
 export const GET = <T>(
   path: string,
   validator?: (data: unknown) => boolean,
-  queryParams?: object
+  queryParams?: object,
+  transformer?: (data: T) => unknown,
 ): Promise<T> =>
   axios
     .get<T>(path, {params: queryParams, paramsSerializer: params => stringify(params)}, )
     .then((response: AxiosResponse<T>) => {
-      if (validator && !validator(response.data)) {
+      const data: any = transformer? transformer(response.data) : response.data;
+      if (validator && !validator(data)) {
         // If a validator is provided, but it fails to validate the provided data...
         throw new Error('Data format mismatch with data received from server.');
       }
 
-      return response.data;
+      return data;
```

Comment:
@PenghaiZhang cast could be avoided by using the type predicate to narrow `unknown` automatically before returning, with:
```suggestion
      if (validator) {
        if (validator(data)) {
          return data;
        }
        throw new Error('Data format mismatch with data received from server.');
      }

      return data;
```

Reply:
this may also need a change to the type for `validator` see https://github.com/openequella/openEQUELLA/pull/1784#pullrequestreview-431975228

Reply:
This function is expected to return a `Promise<T>`. As `data` is `unknown`, `return data` at line 42 may not be OK.

Reply:
That is close to what we want, if the type isn't validated it is `unknown`.

:thinking: the return type may need to be made conditional on the presence of `validator`

Reply:
Okay, we had more of a play with this but it's getting a bit confusing. Perhaps what we've got is good enough for now, and we can try and improve it later. Or if you have time @ChristianMurphy  maybe once this branch is merged in you can have a hack at it then.

Reply:
@edalex-ian @PenghaiZhang https://github.com/PenghaiZhang/openEQUELLA/pull/2

Reply:
Okay, we've had a look but are concerned about how that supports all the different possible permutations. e.g. What If I want a path, no validator, but params; or, path, no validator, maybe params, and a transformer. It's just getting a bit crazy, and we need to merge this one in so as not to block further work.

So let's proceed to merge for now, and then maybe once in your can repoint your PR and show us how all the permutations can be support cleanly... Sorry we can't accommodate just now, but we definitely tried.

Reply:
> but are concerned about how that supports all the different possible permutations. e.g. What If I want a path, no validator, but params; or, path, no validator, maybe params, and a transformer

I'm not sure I follow this, optional params are positional.
in the current implementation
```
(
  path: string,
  validator?: (data: unknown) => boolean,
  queryParams?: object,
  transformer?: (data: T) => unknown,
)
```
The parameters must be added in order (`path`, `path and validator`, `path and validator and query params`, or `path and validator and query params and transformer`).
It isn't possible to skip `validator` and go straight to `transformer` with this syntax.

```ts
export function GET<T>(path: string): Promise<unknown>;
export function GET<T>(
  path: string,
  validator: (data: unknown) => data is T,
  queryParams?: object,
  transformer?: (data: unknown) => T
): Promise<T>;
```

covers all possible combinations.